raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -443,7 +443,7 @@ private void onBecomeLeader(long currentTimeMs) {
     private void flushLeaderLog(LeaderState<T> state, long currentTimeMs) {
         // We update the end offset before flushing so that parked fetches can return sooner.
         updateLeaderEndOffsetAndTimestamp(state, currentTimeMs);
-        log.flush();
+        log.flush(false);
```

Comment:
This is probably ok. The way that kraft manages log retention is a bit different from normal partitions. In general, segments are only deleted once we have a snapshot which covers all of their data. That should mean that there is no risk today of losing the log end offset even if an unflushed empty segment is lost. In other words, even if all of the segments are lost, we should still have a snapshot to derive the log end offset from.

cc @jsancio

Reply:
Yes. Looking at the code, we don't need to have segments to determine the log end offset. When loading the `KafkaMetadataLog`, KRaft `truncateFullyAndStartAt` using the latest snapshot if the log's end offset (`0`) is less than the latest snapshot's end offset. As @hachikuji mentioned, we only delete inactive segments if there is a snapshot that covers the largest offset in the segment.